src/app/qgsattributetabledialog.cpp

```diff
@@ -407,7 +410,26 @@ void QgsAttributeTableDialog::filterColumnChanged( QObject* filterAction )
   mFilterButton->setDefaultAction( qobject_cast<QAction *>( filterAction ) );
   mFilterButton->setPopupMode( QToolButton::InstantPopup );
   mCbxCaseSensitive->setVisible( true );
-  mFilterQuery->setVisible( true );
+  // replace the search line edit with a search widget that is suited to the selected field
+  mFilterQuery->setVisible( false );
+  // delete previous widget
+  if ( mCurrentSearchWidgetWrapper != 0 )
+  {
+    delete mCurrentSearchWidgetWrapper;
+  }
+  QString fieldName = mFilterButton->defaultAction()->text();
+  // get the search widget
+  int fldIdx = mLayer->fieldNameIndex( fieldName );
+    if ( fldIdx < 0 )
+      return;
+  const QString widgetType = mLayer->editorWidgetV2( fldIdx );
+  const QgsEditorWidgetConfig widgetConfig = mLayer->editorWidgetV2Config( fldIdx );
+ //replace with createSearch or so
+ //go to registry and create a create Search method
+  mCurrentSearchWidgetWrapper= QgsEditorWidgetRegistry::instance()->createSearch( widgetType, mLayer, fldIdx, widgetConfig, mFilterContainer);
+  mCurrentSearchWidgetWrapper->widget()->setObjectName("searchy");
+  mCurrentSearchWidgetWrapper->widget()->setVisible( true );
```

Comment:
I tried this, didn't work. I made a function to replace the widgets in the layout and this seems to work.